### ISSUE TITLE
SelectNonNull extension method

### ISSUE THREAD

**Issue by Thaina** [dotnet/corefx#26698]
```C#
public static IEnumerable<V> SelectNonNull<T,V>(this IEnumerable<T> items,Func<T,V> func) where V : class
{
	foreach(var item in items)
	{
		var value = func(item);
		if(value != null)
			yield return value;
	}
}

public static IEnumerable<V> SelectNonNull<T,V>(this IEnumerable<T> items,Func<T,V?> func) where V : struct
{
	foreach(var item in items)
	{
		var value = func(item);
		if(value.HasValue)
			yield return value.Value;
	}
}
```

I don't know what name it should but this extension method is convenient

**Comment by OmarTawfik** [comment]
It would be helpful to follow https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/api-review-process.md

Also, it would be helpful to mention the most common use cases, and why it is specially needed in the  framework, and not as a helper library to a few users who might use it.

**Comment by jnm2** [comment]
This has always struck me as ugly. `.Select(...).NonNull()` is more interesting to me. More generally, it could be `.Select(...).Except(null)` where it's a more efficient single-item version of the existing `Except`.

**Comment by Thaina** [comment]
@jnm2 Interesting. `.ExceptNull()` could be alternative

But `Except(null)` is not possible. Because `Except(null)` will match `Except((IEnumerable)null)`. We need new name for it

And it can't convert nullable to struct, which is the point of this proposal

Also should we `.Select(func).ExceptNull()` or just allow `.ExceptNull(func)` ?

**Comment by stephentoub** [comment]
Why is this not just:
```C#
.Select(func).Where(c => c != null)
```
?

**Comment by jnm2** [comment]
@stephentoub @Thaina seemed particularly interested in the unwrapping of nullable values (second overload).

**Comment by Thaina** [comment]
@stephentoub It obviously shorter and reduce number of delegates

**Comment by stephentoub** [comment]
> It obviously shorter and reduce number of delegates

Yes, that's true for any LINQ usage of multiple operators; you could always create a new operator that represented the combination. I'm not seeing why this particular combination is so valuable and generally in need by many developers that it would warrant a dedicated method.

**Comment by Thaina** [comment]
@stephentoub Because I think this is most common logic. To select something and also filtering at the same time. Without the need to pollute code with `.Where(n => n != null)` everywhere we could just indicate that we don't need `null`

**Comment by Connect-a** [comment]
If we have a `NonNull` (or `FilterOutNull`?), we can convert `IEnumerable<int?>` into `IEnumerable<int>`.
I want to processing list of nullable reference types more easily .

https://docs.microsoft.com/ja-jp/dotnet/csharp/tutorials/nullable-reference-types
